src/target/riscv/riscv-013.c

```diff
@@ -2926,3 +2933,483 @@ void riscv013_clear_abstract_error(struct target *target)
+  COMPLIANCE_TEST(get_field(testvar_read, DMI_ABSTRACTCS_CMDERR) == CMDERR_NOT_SUPPORTED, \
+		  "Illegal COMMAND should result in UNSUPPORTED");
+  dmi_write(target, DMI_ABSTRACTCS, DMI_ABSTRACTCS_CMDERR);
+  dmi_write(target, DMI_COMMAND, 0x55555555);
+  dmi_read(target, &testvar_read, DMI_COMMAND);
+  COMPLIANCE_TEST(testvar_read == 0x55555555, "COMMAND register should be R/W");
+  dmi_read(target, &testvar_read, DMI_ABSTRACTCS);  
+  COMPLIANCE_TEST(get_field(testvar_read, DMI_ABSTRACTCS_CMDERR) == CMDERR_NOT_SUPPORTED, \
+		  "Illegal COMMAND should result in UNSUPPORTED");
+  dmi_write(target, DMI_ABSTRACTCS, DMI_ABSTRACTCS_CMDERR);
+
+  // Basic Abstract Commands
+  uint32_t command = 0;
+  uint32_t busy;
+  command = set_field(command, AC_ACCESS_REGISTER_SIZE, riscv_xlen(target) > 32 ? 3:2);
+  command = set_field(command, AC_ACCESS_REGISTER_TRANSFER, 1);
+  for (unsigned int i = 1 ; i < 32 ; i = i << 1) {
+    command = set_field(command, AC_ACCESS_REGISTER_REGNO, 0x1000 + GDB_REGNO_ZERO + i);
+    command = set_field(command, AC_ACCESS_REGISTER_WRITE, 1);
+    dmi_write(target, DMI_DATA0, i);
+    if (riscv_xlen(target) > 32) {
+      dmi_write(target, DMI_DATA0 + 1, i + 1);
+    }
+    dmi_write(target, DMI_COMMAND, command);
+    do {
+      dmi_read(target, &testvar_read, DMI_ABSTRACTCS);
+      busy = get_field(testvar_read, DMI_ABSTRACTCS_BUSY);
+    } while (busy);
+    dmi_read(target, &testvar_read, DMI_ABSTRACTCS);
+    COMPLIANCE_TEST(get_field(testvar_read, DMI_ABSTRACTCS_CMDERR) == 0, "GPR Writes should be supported.");
+    dmi_write(target, DMI_DATA0, 0xDEADBEEF);
+    if (riscv_xlen(target) > 32) {
+      dmi_write(target, DMI_DATA0 + 1, 0xDEADBEEF);
+    }
+    command = set_field(command, AC_ACCESS_REGISTER_WRITE, 0);
+    dmi_write(target, DMI_COMMAND, command);
+    do {
+      dmi_read(target, &testvar_read, DMI_ABSTRACTCS);
+      busy = get_field(testvar_read, DMI_ABSTRACTCS_BUSY);
+    } while (busy);
+    dmi_read(target, &testvar_read, DMI_ABSTRACTCS);
+    COMPLIANCE_TEST(get_field(testvar_read, DMI_ABSTRACTCS_CMDERR) == 0, "GPR Reads should be supported.");
+    dmi_read(target, &testvar_read, DMI_DATA0);
+    COMPLIANCE_TEST(testvar_read == i, "GPR Reads and writes should be supported.");
+    if (riscv_xlen(target) > 32) {
+      dmi_read(target, &testvar_read, DMI_DATA0 + 1);
+      COMPLIANCE_TEST(testvar_read == (i + 1), "GPR Reads and writes should be supported.");
+    }
+  }
+
+  // ABSTRACTAUTO
+  // See which bits are actually writable
+  dmi_write(target, DMI_ABSTRACTAUTO, 0xFFFFFFFF);
+  uint32_t abstractauto;
+  dmi_read(target, &abstractauto, DMI_ABSTRACTAUTO);
+  dmi_write(target, DMI_ABSTRACTAUTO, 0x0);
+  if (abstractauto > 0) {
+    testvar = 0;
+    // TODO: This mechanism only works when you have a reasonable sized progbuf, which is not
+    // a true compliance requirement.
+    uint32_t result = riscv_set_register(target, GDB_REGNO_S0, 0);
```

Comment:
The "right way" to set S0 when you're not making it visible to the user of the debugger, is to call register_write_direct(), and to read it with register_read_direct(). That makes no difference in your codebase, but it will once the optimizations merge. One of the optimizations is to cache S0 so we're not reading it over and over again, and I think with the functions you use here you may end up reading a cached S0, not the one that was actually modified.

register_set() and register_get() are for use by the user (probably gdb). Since the user can't execute arbitrary programs, they don't handle that case.

(This problem appears numerous times in this function, I just hadn't realized it until I'd read this far.)

Reply:
Can you add some commentary to those functions prototype in a follow-up PR to explain the differences?

Reply:
#241